### PR TITLE
Fix #92

### LINKED ISSUE #92: Magic Module for 'require' missing?
I noticed that this loader generates [AMD magic module](https://github.com/requirejs/requirejs/wiki/Differences-between-the-simplified-CommonJS-wrapper-and-standard-AMD-define#exports) definitions for the "module" and "export" dependencies such that a require statement like the one below succeeds....

~~~~
define("example", ["module", "exports", 'jquery'], function (module, exports,$) { ...
~~~~~

However the Liferay AMD loader doesn't appear to define a 'magic module' for 'require'. I tested this with a couple of examples, and including a module dependency on 'require' results in non-resolution of dependencies. This becomes problematic when using a number of transpilers that generate AMD-compliant code--a good example is TypeScript which generates module definitions that *always* include dependencies on 'require' and 'exports'...

module.ts file
~~~~
export var x=1;
~~~~

running *"tsc -m amd module. ts " produces the following module.js (which is the trouble)

~~~~
define(["require", "exports"], function (require, exports) {
    "use strict";
    exports.x = 1;
});
~~~~

### PR DESCRIPTION
**Pull request by izaera**

### CODE REVIEW

**Comment by ipeychev** [comment]
Hey @izaera,

Thanks for the PR and the efforts!

This PR implements the local **synchronous** require, as described [here](https://github.com/liferay/liferay-amd-loader/issues/92#issuecomment-263141438) (see 2.), doesn't it?:

> Another valid example is when define specifies the dependencies explicitly and there is a local synchronous require, like this:
> ```
> define('sample', ['require', 'a', 'b'], function(require) {
>  var a = require('a');
>  var b = require('b');
> });
> ```

However, there is also asynchronous local require, see [https://github.com/liferay/liferay-amd-loader/issues/92#issuecomment-263141438] (3.):

> ```
> define('sample1', ['require'], function(require) {
>   require(['a', 'b'], function(a, b) {
>   });
> });
> ```

In this case the local require is supposed to load these two modules asynchronous. With the PR, this won't happen, correct?

Thanks,

**Comment by izaera** [comment]
@ipeychev Correct. That's what I mentioned in issue 92:

https://github.com/liferay/liferay-amd-loader/issues/92#issuecomment-288672507

They call it the "Simplified CommonJS Wrapper" scenario in the AMD spec (see https://github.com/requirejs/requirejs/wiki/Differences-between-the-simplified-CommonJS-wrapper-and-standard-AMD-define#sync-require ).

**Comment by izaera** [comment]
See that I created another issue #112 in case we want to implement the "Simplified CommonJS Wrapper" require in the future. However, reading your comments in issue #113 about the rationale behind the loader (https://github.com/liferay/liferay-amd-loader/issues/113#issuecomment-288853467) I don't think it is worth the effort...

**Comment by ipeychev** [comment]
Hmm, no, this is another thing. There are three cases, I will write them in code:
1. Local synchronous require:
```js
define('sample', ['require', 'a', 'b'], function(require) {
 var a = require('a');
 var b = require('b');
});
```
This is what you implemented in this PR.

2. Local asynchronous require:
```js
define('sample1', ['require'], function(require) {
  require(['a', 'b'], function(a, b) {
  });
});
```

This is what won't work now with this PR, because the local require function as written here, doesn't support this.

3. The "Simplified CommonJS Wrapper":
```js
define(function(require, exports, module){
    var foo = require('foo');
    var bar = require('foo/bar');

    exports.doSomething = function(){
        console.log(foo + bar);
    };
});
```
where the loader is supposed to 'scan' the function using `.toString` and load these in advance. In other words, it will load these in advance too, but it is supposed to detect them from the source instead to rely on the developer to specify them. This is for what you created https://github.com/liferay/liferay-amd-loader/issues/112.

I'm talking now about the local **asynchronous** require, not the synchronous one, neither about the "Simplified CommonJS Wrapper".

Thanks,

**Comment by izaera** [comment]
Oh, I see. Sorry. It's true, I only implement the local sync require. 

I was going to say that I can try and implement 2, but after reading the documentation I'm not quite sure if that scenario must be handled... 

In the AMD spec it says:

> This specification defines three special dependency names that have a distinct resolution. If the value of "require", "exports", or "module" appear in the dependency list, the argument should be resolved to the corresponding free variable as defined by the CommonJS modules specification.

And, in CommonJS spec, AFAIK, there's no async require. So I interpret that the AMD spec is not requiring support for async local require. What do you think?

**Comment by izaera** [comment]
As an "off-topic", and regarding 3, I also found this:

> If the dependencies argument is omitted, the module loader MAY choose to scan the factory function for dependencies in the form of require statements

The key word is *MAY*. I guess that in that case we are 100% AMD spec compliant (even if we don't implement #112) because we choose not to scan the function in all cases >:-)

**Comment by ipeychev** [comment]
The local asynchronous require and the "Simplified CommonJS Wrapper" are both supported by RequireJS - this is the only specification people care :) You know, no any user and 99.9999% of the UI dev will never read the spec. They don't care about it. They get the most sophisticated implementation as reference (RequireJS implemented the _MAY_) and will compare everything else with it, in the same way as we do with Chrome and we used to do with IE, when it was the best browser on Earth.
Knowing that, here is my opinion:
1. About "Simplified CommonJS Wrapper", we can forget for now. IMO, this is a sad result from some really bad marijuana the dealers sold to the developers, who invented it :D
2. The local asynchronous require however makes sense. Imagine a module, which loads some modules on demand as a result of user action:
```js
define('main', ['require'], function(require) {
  document.getElementById('button1').addEventListener(function(event) {
      // show loading spinner here and load some modules on demand
      require(['a', 'b'], function(a, b) {
          // do something
      });
  });
});
```
Hope I haven't misunderstood something, but if I haven't, this is a pretty realistic user scenario.

So, my point is that If we leave the local require to support only the synchronous call, we will implement some quite unrealistic use case and break the realistic one.

Thanks,

**Comment by izaera** [comment]
OK. In that case I will implement it too.

It doesn't look very complex: I guess it's just wiring the local require to the global require when it is being invoked async'ly and everything should fit like a sidereal puzzle then ;-)

Should I add more commits to this PR or create a new one? What do you prefer @ipeychev ?

**Comment by izaera** [comment]
BTW, just to add more context, when implementing the local sync require I was trying to mimic what tools like browserify and webpack do when bundling npm modules.

Such tools just copy-paste the npm module directly into the bundle and provide a local sync require to make it work. So, what we are doing with implementing that use case is to allow for someone to bundle npm server modules directly into AMD modules and keep them working correctly (though I agree that it may be very possible that nobody does such thing in the future).

In addition, in our case, we don't need to do that because we don't copy-paste the module verbatim, but transpile and rewrite it so that no local sync requires appear any more.

**Comment by ipeychev** [comment]
I see, okay. And yeah, you may add more commits to this PR.

Thanks!

**Comment by izaera** [comment]
@ipeychev There you are... Those are the commits to implement the async local require. BTW I have also fixed a test fixture which was incorrectly defined.

**Comment by ipeychev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by ipeychev** [comment]
Could you please add in the demo pages examples of usage of the local sync and async require?

**Comment by ipeychev** [comment]
In general, this should do the trick, good job!

File: src/js/script-loader.js [hunk 1]
```diff
@@ -721,17 +721,26 @@ var LoaderProtoMethods = {
                     dependencyImplementations.push(exportsImpl);
                 } else if (dependency === 'require') {
                     var localRequire = function(moduleName) {
-                        var mappedModuleName = configParser.mapModule(moduleName);
+                        var argc = arguments.length;
 
-                        for (var k = 0; k < module.dependencies.length; k++) {
-                            var dependency = module.dependencies[k];
+                        if (argc > 1) {
+                            var error = arguments[argc-1];
```
Comment (izaera) [inline src/js/script-loader.js]: No, they are not used. I will remove them and that will fix the previous issue too :+1:

**Comment by izaera** [comment]
@ipeychev More changes 👆

**Comment by ipeychev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by ipeychev** [comment]
Thank you, pull request merged! [See changes here.](https://github.com/liferay/liferay-amd-loader/compare/ab31c19~2...ab31c19) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request merged**